### PR TITLE
Create namingConvention.md

### PR DESCRIPTION
**Pull request by kivel**
first draft of a file summarising the effort towards a future proof naming of internal variables and EPICS PVs.

### CODE REVIEW

**Comment by anderssandstrom** [comment]
Very good idea to add this doc to ecmccfg to make it traceable.
In general I agree with most of the contents but have a few reflections:
1. For general slave values/errors/.. (on a slave level) maybe we could consider to not have an index, i.e "status" instead of "status00". I just see a risk of software people mixing "status00" as the status of "status01" (starting to count from 0 vs 1).
2. Is your suggestion to also use "value" for binaries ("BI_1" would be "value01")? I think it could be an advantage to have different names on analog values compared to binary when troubleshooting.

Some details (just for the record):
3. For oversampling we need to add one more ID.
4. Guess we should also need to check these "namespaces":
* ax<id>.*
*ds<id>.*
*plcs.plc<id>.*

**Comment by kivel** [comment]
> Very good idea to add this doc to ecmccfg to make it traceable.
> In general I agree with most of the contents but have a few reflections:
> 
> 1. For general slave values/errors/.. (on a slave level) maybe we could consider to not have an index, i.e "status" instead of "status00". I just see a risk of software people mixing "status00" as the status of "status01" (starting to count from 0 vs 1).
> 2. Is your suggestion to also use "value" for binaries ("BI_1" would be "value01")? I think it could be an advantage to have different names on analog values compared to binary when troubleshooting.
do you think it's OK to have 'analog, binary, posAct, ...' or would you prefer to have the direction also in the name, i.e.'binaryInput'
> 
> Some details (just for the record):
> 3. For oversampling we need to add one more ID.
Yes, good point, I was kinda brainstorming directly into the file.

> 4. Guess we should also need to check these "namespaces":
> 
> * ax.*
>   _ds._
>   _plcs.plc._
 certainly, see above.

**Comment by anderssandstrom** [comment]
Direction:
Good point. Today we have BI_x vs BO_x, right. So I think it's a good idea to include also directions. 

Abbreviation or not:
I just reflected on that the "posAct" is an abbreviation.
Should we then use"binIn01", "binOut01", "anaIn01" and "anaOut01". 
or just "bi01", ..., "ao01"?

ec, motion, ds:
For some of the ecmc autogenerated partameter names, we need to update ecmc source. Example:
* "ecmc.appmode" 
* "ecmc.thread.latency.max"
*....

**Comment by kivel** [comment]
> Direction:
> Good point. Today we have BI_x vs BO_x, right. So I think it's a good idea to include also directions.

We _have_ to do this, otherwise the EP2038 gets messy. 

> Abbreviation or not:
> I just reflected on that the "posAct" is an abbreviation.
> Should we then use"binIn01", "binOut01", "anaIn01" and "anaOut01".
> or just "bi01", ..., "ao01"?

Right now, I'm not much of a fan to abbreviate to bi01 or ao03, ...
I'd opt for the analog in/outputs for `analogInput02` and `analogOutput01`, well as `binaryInput03` and `binaryOutput07`. It's more descriptive and not more work, as most configuration are copy-paste anyway. But, I'm open for `analogIn01`, just not `ai06`
Not 100% certain about the position and velocity though. Those, consequently, would have to be `positionActual01` or `velocitySetPoint01`. 
Do we have any restrictions in terms of max char to keep in mind?

> ec, motion, ds:
> For some of the ecmc autogenerated partameter names, we need to update ecmc source. Example:
> 
> * "ecmc.appmode"
> * "ecmc.thread.latency.max"
>   *....

Maybe keep those till the end? Do those have to be individually defined PVs at all, or could they be grouped together into a Epics7 structure? No clue how to do this though, but sounds nice :)

**Comment by anderssandstrom** [comment]
Abbreviation:
Sure, then lets not abbreviate. I'm also ok with the complete "long" name.

length restriction:
The only restriction we have is that the entire name can max be 255 chars, which I guess should be enough.

Epics 7 struct:
Would be cool. I'm not 100% sure how this works. But as I understood it, you can group data that's already in epics. You still need to get the data inside epics somehow. But like I said, i don't really know...

**Comment by kivel** [comment]
EPICS7 struct

see here
https://gist.github.com/kivel/d76792a282d67b4a12310b491b8d3c69

This is a simple example, of course this is all in 'db', no idea how this is done in code. Also, is the API really fixed? The core developers sometimes change the API from one minor to the other. I recall that was the case in 7.0.3 --> 7.0.4, cause some red heads at PSI :)

**Comment by anderssandstrom** [comment]
Nice example of structs. The question is if you can do this from a "hardware" driver like ecmc? Ofcource you could make a pvput but that seems not so optimised. 
BTW, I created this some time ago (one of our integrators requested it): https://github.com/anderssandstrom/e3-ecmcPlugin_Pva
Not sure why its needed, seems safer to take the data via epics db.

**Comment by kivel** [comment]
> Nice example of structs. The question is if you can do this from a "hardware" driver like ecmc? Ofcource you could make a pvput but that seems not so optimised.

Not sure how to do this from hardware. I guess this is possible though. Otherwise, what would be the point, if one would have to create each channel separately and than bundle it together into a struct.
I just received notice that we're _not_ using 'pva' for SLS-2.0!

> BTW, I created this some time ago (one of our integrators requested it): https://github.com/anderssandstrom/e3-ecmcPlugin_Pva
> Not sure why its needed, seems safer to take the data via epics db.

I saw that before, I still can't understand what's the use case, though.

**Comment by kivel** [comment]
@anderssandstrom 

I'm cleaning up a bit in the `db` directories
Any idea what those files are?
db/Beckhoff_6XXX/ecmcEL6224-Analog-chX.template
db/Beckhoff_6XXX/ecmcEL6224-Digital-chX.template

**Comment by anderssandstrom** [comment]
I think they are work in progress for the IO-link terminal. I'm not 100% sure if I made it work in the end. Remember that I hade issues with both TwinCAT and ecmc. Maybe those should not be on master until verified. Can we keep them for now, then I will check when at work next time.

**Comment by anderssandstrom** [comment]
The only nice thing with the PVA plugin is that it's a example of pva usage. I remember it was not easy, so I guess it was probably not the correct way to code it :-)

**Comment by anderssandstrom** [comment]
Hi @kivel 
We now have a slightly modified record naming proposal for ESS:
1. the ecmccfg "inflated" record names will need to start with a "#" (ess calls this "private mode")
2. our property needs to start (after "#") with a char. Suggestion start with "e" as abbreviation for "ec", like used before
3. We now can have more than 20 chars after ":". Not sure yet how many we can have.
Suggestion is basically like these examples:
<P>:#e0s001-BI1
<P>:#e1.s102-Drv02-PosCmd

What do you think?
I guess by adding a lot of MACROS we could handle many different naming scenarios. 
Cheers

**Comment by kivel** [comment]
@anderssandstrom So your naming is set in stone now?

I presume the binaryInput example should be named
":#e0s001-BI01" <-- added "0"

I like it, still needs some getting used to, though 😊
Imho the part after the ":" and before the first "-" should be handled by a dedicated script that sets a variable. Similar to "SCRIPTEXEC", maybe?
By this, we can easily swap the _critical_ section and more easily adapt to the changing environment.
I also have to keep in mind that I have to serve at least 2, more likely 3 facilities at one point. The naming conventions are close, but not necessarily equal.
Good or not? Let me know.
best
Niko

**Comment by anderssandstrom** [comment]
Yes, I meant BI01 ;-)
Good idea with scripts building the parts of the name so it will be more flexible! So then you could define different scripts/function that  build the "#e0s001-" part, right? And then include this in the addSlave.cmd logic. Then we even could have a legacy mode with the HWTYPE included also. Cool!
Not sure here about the standard in macro naming like "P" and "R". Maybe we should have several macros in the name to make it even more generic. I mean we do not need to use the macros. But maybe will look cluttered and hard to follow. Do you have any thoughts on this?
Cheers
A

**Comment by kivel** [comment]
> Yes, I meant BI01 ;-)
> Good idea with scripts building the parts of the name so it will be more flexible! So then you could define different scripts/function that build the "#e0s001-" part, right? And then include this in the addSlave.cmd logic. Then we even could have a legacy mode with the HWTYPE included also. Cool!

That's what I had in mind. :)

> Not sure here about the standard in macro naming like "P" and "R". Maybe we should have several macros in the name to make it even more generic. I mean we do not need to use the macros. But maybe will look cluttered and hard to follow. Do you have any thoughts on this?

I guess it can easily get convoluted if we have a bunch of macros that aren't use. I'd rather have a pile of naming scripts to handle this.

Niko

**Comment by anderssandstrom** [comment]
Hi @kivel
Everythings fine?!

Now I have received internal “formal” approval of starting to rename stuff in ecmccfg according to what we have discussed.

The question is just how we should handle it (it bit hard to start.. so many ways todo it):

**ECMC internal:**
ECMC internal naming is rather well defined in namingconvention.md except the part before the “property”. I would like to keep the ec<id>.s<id> if possible for ecmc internals but maybe this will be confusing. What do you think?!

**Record naming:**
How many macros do you think we should have?
"$(MACRO1)$(MACRO2)<hardcoded_name>"

MACRO1: All the stuff in the beginning.
MACRO2: “e0s001-“ if I remember correct (we can handle with script, like discussed above)
Or more macros?
Should we call them P and R (not really sure what R usually is for).
I guess we need to have a hardcoded property in the end, right? Or any ideas?

**Properties**:
Simple examples for discussion:
Digital inputs      (EL1xxx + more) : **BI01**
Digital outputs   (EL2xxx + more): **BO01**
Analog input      (EL3xxx + more): **AI01**
Analog Outputs (EL4xxx + more): **AO01**
status of slave (all): Stat

**Drives**:
Drv "prefix" for drive
Position setpoint (EL7xxx + more):  **DrvPosCmd01** or **PosSet01**
Velocity setpoint  (EL7xxx + more): **DrvVelCmd01** or **VelSet01**
status of drive: **DrvStat01**
**Encoders**
Enc "prefix" for encoder
Position measurements (EL5xxx + more): **EncPosAct01**
status of encoder: **EncStat01**

**Readbacks**
How to handle read backs: Use **-RB** like today? Or only **RB**:
**VelSet01RB**
Alternative is to make the output records:
 info(asyn:READBACK,"1")


We need also prefix for (atleast):
Reset: **Rst**
Latch: **Ltch**
Enable: **EnaCmd**  ?
Enabled: **EnaAct** ?
Control word: **Cmd** or **Cntrl**
....many more I guess...

Sorry just listed a lot to get discussion going! Sorry for spamming..
What do you think!

**Comment by kivel** [comment]
Hi Anders


ECMC internal is something that not really concerns the naming queens and kings. Still, we should adhere to a common consensus here.


Record naming: I think a we should stick to the omnipresent "${P}foobar" way. Than the script called by 'addSalve' will sort out the "P".

We never use the "R" macro ... well technically I don't use "P" either in my normal IOCs 😊

I'm happy to discuss this, if you feel that more macros will help.


Properties:

Yes, BI, BO, AI, AO with two digits in hex (I just got my C9900-M993, which has 11 BI). Look much cleaner in hex 😊

Stat, Online, Operational, AlStatInit, ...


Drives:

I like VelSet and PosAct better than DrvPosCmd


Encoders:

The same, just PosAct.

Do you thinks it's beneficial to know whether the position stems from an encoder or open-loop counter in the drive?


Following this logic, the status of drive and Encoders, should be "StatEnc" and "StatDrv". If we decide for "DrvPosCmd" the status should be "DrvStat"


Readback:

My boss says, -RB. Has to do with tracing via caPutLog and Archiver stuff.

We could have a global Macro for "-RB" vs "RB". I prefer -RB.


"

Enable: EnaCmd ?
Ena
Enabled: EnaAct ?
Ena-RB?
Control word: Cmd or Cntrl
Crtl :)
"


We shouldn't restrict ourselves. For special devices, we need the room to do special solutions.

i.a. the C9900. It's nice to have "pc11173:ec0-s7-C9900-M993-BO1_GREEN", but "-Btn01R" would also do.


Also,

ESTOP

INHIBIT, ... need some standard I guess.


Should we meet via video chat, I guess it's more productive that writing long emails?


best

Niko

--
Paul Scherrer Institute
Dr. Niko Kivel
Forschungsstrasse 111
WBGB/017
5232 Villigen PSI
Switzerland

phone: +41 56 310 2226
email: niko.kivel@psi.ch
________________________________

**Comment by anderssandstrom** [comment]
Hi Niko,

Yes a video chat would be nice!
I just write some quick comments here!

**Macros**:
I think your proposal in rc branch look great. P is enough, the we can build as we want in the script.
The classical way will still use the new properties I guess, so it will not be completed backward compatible, but that is fine!

**"Enc", "Drv":**
Both encoder and drive have control and status words, so then I think we need the prefix "Enc" or "Drv" so maybe we should keep it also for (Enc)PosAct and (Drv)VelSet also?

**Ena, Ena-RB, EnaAct:**
These are three different values:
Ena: The commands
Ena-RB:  Readback of command
EnaAct: Is the drive enabled.

**-RB:**
Good info!

Cheers
Anders
